File: tests/test_hallucination_detector_v2.py
"""
Tests for Core/hallucination_detector_v2.py — AdvancedHallucinationDetector

Tests cover:
  - HallucinationReport dataclass behaviour
  - ClaimCheck dataclass
  - detect_hallucinations() API
  - Legacy detect() wrapper
  - to_legacy_dict() format for UI compat
  - Edge cases (empty input, no evidence)
"""

import pytest
from unittest.mock import MagicMock


# ─── Fixtures ──────────────────────────────────────────────────────────────


@pytest.fixture
def detector():
    """Create detector with mocked CrossEncoder to avoid downloads."""
    from Core.hallucination_detector_v2 import AdvancedHallucinationDetector

    d = AdvancedHallucinationDetector.__new__(AdvancedHallucinationDetector)
    d.nli_model = None  # No real NLI model
    d.nli_available = False
    d.logger = MagicMock()
    return d


@pytest.fixture
def grounded_answer():
    return "Python was created by Guido van Rossum and released in 1991."


@pytest.fixture
def evidence():
    return [
        "Python is a programming language created by Guido van Rossum. It was first released in 1991.",
        "Python supports multiple programming paradigms including OOP and functional.",
    ]


# ─── Import ────────────────────────────────────────────────────────────────


def test_import():
    from Core.hallucination_detector_v2 import (
        AdvancedHallucinationDetector,
        ClaimCheck,
        HallucinationReport,
    )

    assert AdvancedHallucinationDetector is not None
    assert ClaimCheck is not None
    assert HallucinationReport is not None


# ─── ClaimCheck dataclass ──────────────────────────────────────────────────


def test_claim_check_creation():
    from Core.hallucination_detector_v2 import ClaimCheck

    c = ClaimCheck(
        claim_text="Python was created in 1991",
        hallucination_type=None,
        confidence=0.95,
        evidence_snippet="first released in 1991",
    )
    assert c.claim_text == "Python was created in 1991"
    assert c.hallucination_type is None
    assert c.confidence == 0.95


def test_claim_check_flagged():
    from Core.hallucination_detector_v2 import ClaimCheck

    c = ClaimCheck(
        claim_text="Python was released in 2020",
        hallucination_type="numerical_error",
        confidence=0.8,
    )
    assert c.hallucination_type == "numerical_error"


# ─── HallucinationReport ──────────────────────────────────────────────────


def test_report_properties():
    from Core.hallucination_detector_v2 import ClaimCheck, HallucinationReport

    claims = [
        ClaimCheck("Claim A", None, 0.9),
        ClaimCheck("Claim B", "ungrounded", 0.7),
        ClaimCheck("Claim C", "numerical_error", 0.8),
    ]
    report = HallucinationReport(
        claims=claims,
        probability=0.35,
        by_type={"ungrounded": ["Claim B"], "numerical_error": ["Claim C"]},
        summary="2 issues found",
    )
    assert report.total_claims == 3
    assert report.hallucinated_claims == 2
    assert report.has_hallucinations is True
    assert report.hallucination_rate == 0.35


def test_report_no_hallucinations():
    from Core.hallucination_detector_v2 import ClaimCheck, HallucinationReport

    claims = [
        ClaimCheck("Good claim", None, 0.95),
    ]
    report = HallucinationReport(claims=claims, probability=0.0, by_type={}, summary="Clean")
    assert not report.has_hallucinations
    assert report.hallucinated_claims == 0


def test_to_legacy_dict():
    from Core.hallucination_detector_v2 import ClaimCheck, HallucinationReport

    claims = [
        ClaimCheck("Good", None, 0.9),
        ClaimCheck("Contradicted", "nli_contradiction", 0.85),
        ClaimCheck("Unsupported", "ungrounded", 0.7),
    ]
    report = HallucinationReport(
        claims=claims,
        probability=0.40,
        by_type={"nli_contradiction": ["Contradicted"], "ungrounded": ["Unsupported"]},
        summary="Issues",
    )
    legacy = report.to_legacy_dict()
    assert legacy["has_hallucinations"] is True
    assert legacy["hallucination_rate"] == 0.40
    assert legacy["total_claims"] == 3
    assert len(legacy["refuted_claims"]) == 1  # nli_contradiction
    assert len(legacy["unsupported_claims"]) == 1  # ungrounded
    assert "summary" in legacy


# ─── detect_hallucinations() ──────────────────────────────────────────────


def test_detect_hallucinations_returns_report(detector, grounded_answer, evidence):
    """Main API should return a HallucinationReport."""
    from Core.hallucination_detector_v2 import HallucinationReport

    report = detector.detect_hallucinations(grounded_answer, evidence)
    assert isinstance(report, HallucinationReport)
    assert 0.0 <= report.probability <= 1.0
    assert isinstance(report.claims, list)


def test_detect_hallucinations_low_prob_for_grounded(detector, evidence):
    """Well-grounded answer should get low probability."""
    answer = "Python supports multiple programming paradigms including OOP."
    report = detector.detect_hallucinations(answer, evidence)
    # Without NLI model, we rely on heuristic checks
    assert isinstance(report.probability, float)


def test_detect_hallucinations_empty_answer(detector, evidence):
    """Empty answer should be handled gracefully."""
    report = detector.detect_hallucinations("", evidence)
    assert isinstance(report.probability, float)


def test_detect_hallucinations_empty_evidence(detector, grounded_answer):
    """Empty evidence list should not crash."""
    report = detector.detect_hallucinations(grounded_answer, [])
    assert isinstance(report.probability, float)


# ─── Legacy detect() wrapper ──────────────────────────────────────────────


def test_detect_legacy_returns_dict(detector, grounded_answer, evidence):
    """Legacy detect() should return dict with expected keys."""
    result = detector.detect(grounded_answer, evidence)
    assert isinstance(result, dict)
    assert "has_hallucinations" in result
    assert "hallucination_rate" in result
    assert "total_claims" in result


# ─── Edge Cases ────────────────────────────────────────────────────────────


def test_very_long_answer(detector):
    """Very long answer should not crash."""
    long_answer = "This is a sentence about science. " * 200
    evidence = ["Science is important."]
    report = detector.detect_hallucinations(long_answer, evidence)
    assert isinstance(report.probability, float)


def test_single_word_answer(detector):
    """Single word answer should work."""
    report = detector.detect_hallucinations("Yes", ["Some evidence text here."])
    assert isinstance(report.probability, float)

File: rust_output/src/test_hallucination_detector_v2.rs
/// Tests for Core/hallucination_detector_v2::py — AdvancedHallucinationDetector
/// 
/// Tests cover:
/// - HallucinationReport dataclass behaviour
/// - ClaimCheck dataclass
/// - detect_hallucinations() API
/// - Legacy detect() wrapper
/// - to_legacy_dict() format for UI compat
/// - Edge cases (empty input, no evidence)

use anyhow::{Result, Context};
use std::collections::HashMap;

/// Create detector with mocked CrossEncoder to avoid downloads.
pub fn detector() -> () {
    // Create detector with mocked CrossEncoder to avoid downloads.
    // TODO: from Core.hallucination_detector_v2 import AdvancedHallucinationDetector
    let mut d = AdvancedHallucinationDetector.__new__(AdvancedHallucinationDetector);
    d.nli_model = None;
    d.nli_available = false;
    d.logger = MagicMock();
    d
}

pub fn grounded_answer() -> () {
    "Python was created by Guido van Rossum and released in 1991.".to_string()
}

pub fn evidence() -> () {
    vec!["Python is a programming language created by Guido van Rossum. It was first released in 1991.".to_string(), "Python supports multiple programming paradigms including OOP and functional.".to_string()]
}

pub fn test_import() -> () {
    // TODO: from Core.hallucination_detector_v2 import AdvancedHallucinationDetector, ClaimCheck, HallucinationReport
    assert!(AdvancedHallucinationDetector.is_some());
    assert!(ClaimCheck.is_some());
    assert!(HallucinationReport.is_some());
}

pub fn test_claim_check_creation() -> () {
    // TODO: from Core.hallucination_detector_v2 import ClaimCheck
    let mut c = ClaimCheck(/* claim_text= */ "Python was created in 1991".to_string(), /* hallucination_type= */ None, /* confidence= */ 0.95_f64, /* evidence_snippet= */ "first released in 1991".to_string());
    assert!(c.claim_text == "Python was created in 1991".to_string());
    assert!(c.hallucination_type.is_none());
    assert!(c.confidence == 0.95_f64);
}

pub fn test_claim_check_flagged() -> () {
    // TODO: from Core.hallucination_detector_v2 import ClaimCheck
    let mut c = ClaimCheck(/* claim_text= */ "Python was released in 2020".to_string(), /* hallucination_type= */ "numerical_error".to_string(), /* confidence= */ 0.8_f64);
    assert!(c.hallucination_type == "numerical_error".to_string());
}

pub fn test_report_properties() -> () {
    // TODO: from Core.hallucination_detector_v2 import ClaimCheck, HallucinationReport
    let mut claims = vec![ClaimCheck("Claim A".to_string(), None, 0.9_f64), ClaimCheck("Claim B".to_string(), "ungrounded".to_string(), 0.7_f64), ClaimCheck("Claim C".to_string(), "numerical_error".to_string(), 0.8_f64)];
    let mut report = HallucinationReport(/* claims= */ claims, /* probability= */ 0.35_f64, /* by_type= */ HashMap::from([("ungrounded".to_string(), vec!["Claim B".to_string()]), ("numerical_error".to_string(), vec!["Claim C".to_string()])]), /* summary= */ "2 issues found".to_string());
    assert!(report.total_claims == 3);
    assert!(report.hallucinated_claims == 2);
    assert!(report.has_hallucinations == true);
    assert!(report.hallucination_rate == 0.35_f64);
}

pub fn test_report_no_hallucinations() -> () {
    // TODO: from Core.hallucination_detector_v2 import ClaimCheck, HallucinationReport
    let mut claims = vec![ClaimCheck("Good claim".to_string(), None, 0.95_f64)];
    let mut report = HallucinationReport(/* claims= */ claims, /* probability= */ 0.0_f64, /* by_type= */ HashMap::new(), /* summary= */ "Clean".to_string());
    assert!(!report.has_hallucinations);
    assert!(report.hallucinated_claims == 0);
}

pub fn test_to_legacy_dict() -> () {
    // TODO: from Core.hallucination_detector_v2 import ClaimCheck, HallucinationReport
    let mut claims = vec![ClaimCheck("Good".to_string(), None, 0.9_f64), ClaimCheck("Contradicted".to_string(), "nli_contradiction".to_string(), 0.85_f64), ClaimCheck("Unsupported".to_string(), "ungrounded".to_string(), 0.7_f64)];
    let mut report = HallucinationReport(/* claims= */ claims, /* probability= */ 0.4_f64, /* by_type= */ HashMap::from([("nli_contradiction".to_string(), vec!["Contradicted".to_string()]), ("ungrounded".to_string(), vec!["Unsupported".to_string()])]), /* summary= */ "Issues".to_string());
    let mut legacy = report.to_legacy_dict();
    assert!(legacy["has_hallucinations".to_string()] == true);
    assert!(legacy["hallucination_rate".to_string()] == 0.4_f64);
    assert!(legacy["total_claims".to_string()] == 3);
    assert!(legacy["refuted_claims".to_string()].len() == 1);
    assert!(legacy["unsupported_claims".to_string()].len() == 1);
    assert!(legacy.contains(&"summary".to_string()));
}

/// Main API should return a HallucinationReport.
pub fn test_detect_hallucinations_returns_report(detector: String, grounded_answer: String, evidence: String) -> () {
    // Main API should return a HallucinationReport.
    // TODO: from Core.hallucination_detector_v2 import HallucinationReport
    let mut report = detector.detect_hallucinations(grounded_answer, evidence);
    assert!(/* /* isinstance(report, HallucinationReport) */ */ true);
    assert!((0.0_f64 <= report.probability) && (report.probability <= 1.0_f64));
    assert!(/* /* isinstance(report.claims, list) */ */ true);
}

/// Well-grounded answer should get low probability.
pub fn test_detect_hallucinations_low_prob_for_grounded(detector: String, evidence: String) -> () {
    // Well-grounded answer should get low probability.
    let mut answer = "Python supports multiple programming paradigms including OOP.".to_string();
    let mut report = detector.detect_hallucinations(answer, evidence);
    assert!(/* /* isinstance(report.probability, float) */ */ true);
}

/// Empty answer should be handled gracefully.
pub fn test_detect_hallucinations_empty_answer(detector: String, evidence: String) -> () {
    // Empty answer should be handled gracefully.
    let mut report = detector.detect_hallucinations("".to_string(), evidence);
    assert!(/* /* isinstance(report.probability, float) */ */ true);
}

/// Empty evidence list should not crash.
pub fn test_detect_hallucinations_empty_evidence(detector: String, grounded_answer: String) -> () {
    // Empty evidence list should not crash.
    let mut report = detector.detect_hallucinations(grounded_answer, vec![]);
    assert!(/* /* isinstance(report.probability, float) */ */ true);
}

/// Legacy detect() should return dict with expected keys.
pub fn test_detect_legacy_returns_dict(detector: String, grounded_answer: String, evidence: String) -> () {
    // Legacy detect() should return dict with expected keys.
    let mut result = detector.detect(grounded_answer, evidence);
    assert!(/* /* isinstance(result, dict) */ */ true);
    assert!(result.contains(&"has_hallucinations".to_string()));
    assert!(result.contains(&"hallucination_rate".to_string()));
    assert!(result.contains(&"total_claims".to_string()));
}

/// Very long answer should not crash.
pub fn test_very_long_answer(detector: String) -> () {
    // Very long answer should not crash.
    let mut long_answer = ("This is a sentence about science. ".to_string() * 200);
    let mut evidence = vec!["Science is important.".to_string()];
    let mut report = detector.detect_hallucinations(long_answer, evidence);
    assert!(/* /* isinstance(report.probability, float) */ */ true);
}

/// Single word answer should work.
pub fn test_single_word_answer(detector: String) -> () {
    // Single word answer should work.
    let mut report = detector.detect_hallucinations("Yes".to_string(), vec!["Some evidence text here.".to_string()]);
    assert!(/* /* isinstance(report.probability, float) */ */ true);
}
